src/runtime/internal/database.server.ts

```diff
@@ -113,20 +115,63 @@ async function _checkAndImportDatabaseIntegrity(event: H3Event, collection: stri
 
     // Delete old version -- checksum exists but does not match with bundled checksum
     await db.exec(`DELETE FROM ${tables.info} WHERE id = ?`, [`checksum_${collection}`])
+
+    if (!unchangedStructure) {
+      // we need to drop the table and recreate it
+      await db.exec(`DROP TABLE IF EXISTS ${tables[collection]}`)
+    }
   }
 
   const dump = await loadDatabaseDump(event, collection).then(decompressSQLDump)
+  let hashesInDb: string[] = []
+
+  if (unchangedStructure) {
+    // get the list of hash to insert
+    const hashListFromTheDump = new Set(dump.map(row => row.split(' -- ').pop()).filter(Boolean))
+
+    // get the list of hash in the database
+    const hashesInDbRecords = await db.all<{ __hash__: string }>(`SELECT __hash__ FROM ${tables[collection]}`).catch(() => [] as { __hash__: string }[])
+    hashesInDb = hashesInDbRecords.map(r => r.__hash__)
+
+    // get the list of hash to delete
+    const hashesToDelete = hashesInDb.filter(hash => !hashListFromTheDump.has(hash))
+    if (hashesToDelete.length) {
+      await db.exec(`DELETE FROM ${tables[collection]} WHERE __hash__ IN (${hashesToDelete.map(() => '?').join(',')})`, hashesToDelete)
+    }
+  }
 
   await dump.reduce(async (prev: Promise<void>, sql: string) => {
     await prev
-    await db.exec(sql).catch((err: Error) => {
+
+    // in D1, there is a bug where semicolons and comments can't work together
+    // so we need to split the SQL and remove the comment
+    // @see https://github.com/cloudflare/workers-sdk/issues/3892
+    const [statement, hash] = sql.split(' -- ')
```

Comment:
This will fail if query contains ` -- `. Imagine if the content contains `sql` snippet.

we can use `hashListFromTheDump` array to find line hash value and remove via `substring`

Something like

```ts
const hash = hashListFromTheDump[index]
const statement = sql.substring(0, sql.length - ('  -- ' + hash).length)
```

Reply:
good catch yeh !